src/node_id.rs

```diff
@@ -0,0 +1,282 @@
+            return Err(DistanceIsZero);
+        } else {
+            return Ok(metric_node);
+        }
+    }
+
+    fn metric_distance(&self, other: &NodeId) -> Result<U256, DistanceIsZero> {
+        let a = U256::from(self.discohash.clone().as_slice());
+        let b = U256::from(other.discohash.clone().as_slice());
+        // xor
+        let distance = a ^ b;
+        if distance == U256::from(0) {
+            return Err(DistanceIsZero);
+        } else {
+            return Ok(distance);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{KadMetric, NodeId};
+    use crate::ed25519::Keypair;
+    use crate::error::{DistanceIsZero, ParseError};
+    use disco::hash;
+    use rand;
+
+    // TODO: change when error becomes more descriptive
+    // for the moment, all ParseError == ParseError
+    impl PartialEq for ParseError {
+        fn eq(&self, other: &ParseError) -> bool {
+            true
+        }
+    }
+
+    // \A DistanceIsZero == DistanceIsZero
+    impl PartialEq for DistanceIsZero {
+        fn eq(&self, other: &DistanceIsZero) -> bool {
+            true
+        }
+    }
+
+    #[test]
+    fn node_id_is_public_key() {
+        let key = Keypair::generate(&mut rand::thread_rng());
+        let keyhash = hash(key.public.as_bytes(), 32);
+        let node_id = NodeId { discohash: keyhash };
+        assert!(node_id.is_public_key(key.public));
+    }
+
+    #[test]
+    fn distance_works() {
+        let node_id = NodeId::generate();
+        let clone_node_id = node_id.clone();
+        let distance = &node_id.distance(&clone_node_id);
+        let new_node_id = NodeId::generate();
+        let distance2 = &node_id.distance(&new_node_id);
+        assert_eq!(distance.as_ref().unwrap_err(), &DistanceIsZero);
+        // assert!(distance2.as_ref().unwrap() != &DistanceIsZero); // if uncommented, compiler error below generated `=>` ok I guess
+        //                                      ^^ no implementation for `node_id::NodeId == error::DistanceIsZero`
+    }
+
+    #[test]
+    fn metric_distance_works() {
+        let node_id = NodeId::generate();
+        let clone_node_id = node_id.clone();
+        let distance = &node_id.metric_distance(&clone_node_id);
+        let new_node_id = NodeId::generate();
+        let distance2 = &node_id.distance(&new_node_id);
+        assert_eq!(distance.as_ref().unwrap_err(), &DistanceIsZero);
+        // assert!(distance2.as_ref().unwrap() != &DistanceIsZero); // if uncommented, compiler error below generated `=>` ok I guess
+        //                                      ^^ no implementation for `node_id::NodeId == error::DistanceIsZero`
+    }
+
+    #[test]
+    fn to_base58_then_back() {
+        let node_id = NodeId::generate();
+        let second: NodeId = node_id.to_base58().parse().unwrap();
+        assert_eq!(node_id, second);
+    }
+
+    #[test]
+    fn incorrect_length_yields_parse_error() {
+        let mut big_data = vec![1u8; 33];
```

Comment:
these don't have to be mut

Reply:
true, I just changed it but it is just a test